server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -1245,8 +1262,15 @@ private void checkForTimeout(long requestId) {
         }
     }
 
+    @Override
+    public void onConnectionOpened(Transport.Connection connection) {
+        transportVersions.put(connection.getNode(), connection.getTransportVersion());
```

Comment:
This seems a little clunky to maintain this mapping here based on the versions that have been seen *by this node*. It also means every node would be keeping this map. It's not a lot, but doesn't seem necessary.

Could we instead adjust DiscoveryNode to keep TransportVersion in addition to Version, so that a similar pattern can be used as with the existing version barrier? @DaveCTurner thoughts?

Reply:
> Could we instead adjust DiscoveryNode to keep TransportVersion in addition to Version

I agree that we will have to keep track of `TransportVersion#CURRENT` in the cluster state itself for every node in the cluster.

However, I'd rather we didn't add this field to `DiscoveryNode`, because that makes it very accessible to a lot of places that have no business looking at it. I worry that this introduces a risk that we will start making decisions incorrectly based on `DiscoveryNode#transportVersion` rather than obtaining the transport version from the specific connection to be used for communication.

Instead I'd prefer to introduce a separate structure in the cluster state which tracks the node/transport-version mapping.

Reply:
So alongside `DiscoveryNodes`, have a new set of classes that handles the TransportVersion side of things?

Reply:
I'm not sure about "set of classes", I was expecting this just to be a `Map<String,TransportVersion>`.

Reply:
I've added that to the cluster state. Is that the sort of approach you mean? Where else do I need to make changes to propagate this properly?